src/dataspaces.jl

```diff
@@ -121,51 +121,165 @@ function isnull(dspace::Dataspace)
+
+Convert `obj` to a `BlockRange` object.
+
+# External links
+- [HDF5 User Guide, section 7.4.2.1 "Selecting Hyperslabs"](https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDataspaces%2FHDF5_Dataspaces_and_Partial_I_O.htm%23TOC_7_4_2_Programming_Modelbc-8&rhtocid=7.2.0_2)
+"""
+function BlockRange(; start::Integer, stride::Integer=1, count::Integer=1, block::Integer=1)
+    if count == -1
+        count = API.H5S_UNLIMITED
+    end
+    BlockRange(start - 1, stride, count, block)
+end
+BlockRange(start::Integer; stride=1, count=1, block=1) =
+    BlockRange(; start=start, stride=stride, count=count, block=block)
+BlockRange(r::AbstractUnitRange; stride=max(length(r), 1), count=1) =
+    BlockRange(; start=first(r), stride=stride, count=count, block=length(r))
+BlockRange(r::OrdinalRange) = BlockRange(; start=first(r), stride=step(r), count=length(r))
+BlockRange(br::BlockRange) = br
+
+Base.to_index(d::Dataset, br::BlockRange) = br
+Base.length(br::BlockRange) = Int(br.count * br.block)
+
+function Base.show(io::IO, br::BlockRange)
+    start = Int(br.start0 + 1)
+    # choose the simplest possible representation
+    if br.count == 1
+        if br.block == 1
+            # integer
+            print(io, start)
         else
-            error("index must be range or integer")
+            # UnitRange
+            print(io, range(start; length=Int(br.block)))
+        end
+    elseif br.block == 1 && br.count != API.H5S_UNLIMITED
+        # StepRange
+        print(io, range(start; step=Int(br.stride), length=Int(br.count)))
+    else
+        # BlockRange(int; ...)
+        print(io, BlockRange, "(start=", start)
+        if br.stride != 1
+            print(io, ", stride=", Int(br.stride))
+        end
+        if br.count != 1
+            print(io, ", count=", br.count == API.API.H5S_UNLIMITED ? -1 : Int(br.count))
         end
-        if dsel_start[k] < 0 ||
-            dsel_start[k] + (dsel_count[k] - 1) * dsel_stride[k] >= dims[n_dims - k + 1]
-            println(dsel_start)
-            println(dsel_stride)
-            println(dsel_count)
-            println(reverse(dims))
-            error("index out of range")
+        if br.block != 1
+            print(io, ", block=", Int(br.block))
         end
+        print(io, ")")
     end
-    API.h5s_select_hyperslab(
-        dsel_id, API.H5S_SELECT_SET, dsel_start, dsel_stride, dsel_count, C_NULL
-    )
-    return Dataspace(dsel_id)
+end
+
+function Base.range(br::BlockRange)
+    start = Int(br.start0 + 1)
+    if br.count == 1
+        # UnitRange
+        return range(start; length=Int(br.block))
+    elseif br.block == 1 && br.count != API.H5S_UNLIMITED
+        # StepRange
+        return range(start; step=Int(br.stride), length=Int(br.count))
+    else
+        error("$br cannot be converted to a Julia range")
+    end
+end
+Base.convert(::Type{AbstractRange}, br::BlockRange) = range(br)
```

Comment:
```suggestion
Base.convert(::Type{T <: AbstractRange}, br::BlockRange)::T where T = range(br)
```

This would allow `Base.convert(UnitRange, ...)` and `Base.convert(StepRange, ...)` as well. Could we also add a few conversion tests?

Reply:
i'm not sure that will work generally?

Reply:
or i'm not sure what you had in mind.

Reply:
Currently `convert(AbstractRange, BlockRange(0, 2, 2, 1))` works. I would also like `convert(StepRange, BlockRange(0, 2, 2, 1))` or `convert(UnitRange, BlockRange(0, 1, 1, 10))` to work.

This would allow something like the following to work.
```
julia> ranges = OrdinalRange[]
OrdinalRange[]

julia> push!(ranges, BlockRange(0, 2, 2, 1))
1-element Vector{OrdinalRange}:
 1:2:3

julia> push!(ranges, BlockRange(0, 1, 1, 10))
2-element Vector{OrdinalRange}:
 1:2:3
 1:10
```

Meanwhile, this should fail:

```julia
julia> convert(UnitRange, BlockRange(0, 2, 2, 1))
ERROR: TypeError: in typeassert, expected UnitRange, got a value of type StepRange{Int64, Int64}
Stacktrace:
 [1] convert(#unused#::Type{UnitRange}, br::BlockRange)
   @ Main .\REPL[22]:1
 [2] top-level scope
   @ REPL[51]:1
```